### PR TITLE
docs(material/table): additional details on table data sources

### LINKED ISSUE #21887: feat(COMPONENT): Make DataSource<T> an interface
#### Feature Description
Instead of an entirely abstract class (see https://github.com/angular/components/blob/master/src/cdk/collections/data-source.ts), why not make it an interface? 

#### Use Case
This would enable users such as myself to create custom data sources that extend _other_ classes. With DataSource<T> being an abstract class, it means I have to extend that to make a custom DataSource and therefore cannot extend any other class with my custom DataSource. My use case is that i have a logging base class, and want to extend it in a new class to also act as a custom Data Source.

### PR DESCRIPTION
**Pull request by s4m0r4m4**
docs(table): additional details on table data sources

Provides additional documentation and clarification on extending the base DataSource class with custom data sources either by extending or implementing.

Fixes #21887

### CODE REVIEW

**Comment by s4m0r4m4** [comment]
@crisbeto - here's the PR we discussed. Feel to propose or modify anything!

File: src/material/table/table.md [hunk 1]
```diff
@@ -107,13 +107,22 @@ automatically trigger an update to the rows each time a new data array is emitte
...
-right time to clean up any subscriptions that may have been registered during the connect process.
+A DataSource is simply a class that has at a minimum the following methods: `connect` and 
+`disconnect`. The `connect` method will be called by the table to provide an Observable that emits 
+the data array that should be rendered. The table will call `disconnect` when the table is destroyed,
+which may be the right time to clean up any subscriptions that may have been registered in the 
+`connect` method. 
+
+Although Angular Material provides a ready-made table DataSource class, `MatTableDataSource`, you may
+want to create your own custom DataSource class for more complex use cases. This can be done by 
+`extending` the abstract DataSource class with a custom DataSource class that then implements the
```
Comment (crisbeto) [inline src/material/table/table.md:121]: I don't think you need the backticks around "extending" here.

**Review by crisbeto — COMMENTED** [review]
LGTM with one nit. You'll also have to change the commit message title to be `docs(material/table)` instead of `docs(table)` in order for the lint check to pass.

**Review by crisbeto — APPROVED** [review]

**Comment by s4m0r4m4** [comment]
Ok I've addressed those two issues you mentioned, let me know if there's anything else!

**Pull request merged**